### ISSUE TITLE
训练dataset_demo过程中指标异常问题

### ISSUE THREAD

**Issue by ZHANG-GangHan** [hukaixuan19970627/yolov5_obb#441]
直接从仓库中拉取代码，并用以下命令直接训练
`python train.py --weights weights/yolov5x.pt --data data/yolov5obb_demo.yaml --hyp data/hyps/obb/hyp.finetune_dota.yaml --epochs 10 --batch-size 1 --img 1024 --device 0`
训练过程中loss下降，但验证的时候指标均为0，并且显示labels也为0

**Comment by ZHANG-GangHan** [comment]
Starting training for 10 epochs...
I0817 20:47:22.107875 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       0/9     3.02G    0.1152    0.1489   0.07375    0.2015        56      1024: 100%|██████████| 1/1 [00:04<00:00,  4.28s/it]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  5.09it/s]
I0817 20:47:26.739942 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:47:39.551578 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       1/9     5.37G    0.1151    0.1437   0.07382    0.2038        61      1024: 100%|██████████| 1/1 [00:00<00:00,  1.14it/s]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  5.00it/s]
I0817 20:47:40.655594 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:47:53.786352 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       2/9     5.37G     0.117    0.1528   0.07368    0.2034        88      1024: 100%|██████████| 1/1 [00:00<00:00,  1.21it/s]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  5.07it/s]
I0817 20:47:54.834279 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:48:06.674522 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       3/9     5.37G     0.117     0.131   0.07235    0.2027        56      1024: 100%|██████████| 1/1 [00:00<00:00,  1.71it/s]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  5.14it/s]
I0817 20:48:07.483672 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:48:19.459146 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       4/9     5.37G    0.1166    0.2747   0.07333    0.2019       146      1024: 100%|██████████| 1/1 [00:01<00:00,  1.19s/it]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  4.89it/s]
I0817 20:48:20.883066 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:48:33.056436 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       5/9     5.37G    0.1154    0.1469   0.07381    0.2022        91      1024: 100%|██████████| 1/1 [00:00<00:00,  1.23it/s]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  4.83it/s]
I0817 20:48:34.106064 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:48:46.260023 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       6/9     5.37G    0.1169    0.1401   0.07303    0.2051        80      1024: 100%|██████████| 1/1 [00:00<00:00,  1.24it/s]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  4.97it/s]
I0817 20:48:47.290685 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:48:59.454138 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       7/9     5.37G    0.1135    0.1593   0.07121    0.2005        49      1024: 100%|██████████| 1/1 [00:00<00:00,  1.21it/s]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  5.06it/s]
I0817 20:49:00.500838 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:49:12.472740 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       8/9     5.37G    0.1181    0.1347   0.07118    0.2023        61      1024: 100%|██████████| 1/1 [00:00<00:00,  1.21it/s]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  5.19it/s]
I0817 20:49:13.505734 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:49:25.418127 140365688276800 train.py:293]
     Epoch   gpu_mem       box       obj       cls     theta    labels  img_size
       9/9     5.37G     0.116    0.1308   0.07353    0.2024        54      1024: 100%|██████████| 1/1 [00:00<00:00,  1.69it/s]
               Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  5.17it/s]
I0817 20:49:26.217280 140365688276800 val.py:280]                  all          1          0          0          0          0          0
I0817 20:49:38.241526 140365688276800 train.py:419]
10 epochs completed in 0.038 hours.
Optimizer stripped from runs/train/exp/weights/last.pt, 178.0MB
Optimizer stripped from runs/train/exp/weights/best.pt, 178.0MB
I0817 20:49:41.719453 140365688276800 train.py:424]
Validating runs/train/exp/weights/best.pt...
I0817 20:49:41.903907 140365688276800 yolo.py:247] Fusing layers...
I0817 20:49:42.454034 140365688276800 torch_utils.py:236] Model Summary: 476 layers, 88510849 parameters, 0 gradients
               
Class     Images     Labels          P          R  HBBmAP@.5  HBBmAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  4.87it/s]
I0817 20:49:42.676683 140365688276800 val.py:280]                  all          1          0          0          0          0          0

**Comment by hukaixuan19970627** [comment]
如果使用demo文件的话，因为数据量极少，训练时的各种指标为0是正常现象，demo只是让人快速上手整个训练流程，你可以换个大一点的数据集试试看，比如DOTA

**Comment by ZHANG-GangHan** [comment]
但labels那边为什么也是打印出0啊？这个位置应该是打印样本中的标签数量的，应该不会受到数据量的影响才对

**Comment by HaoyuScut** [comment]
你好，请问解决了吗，我也遇到该问题了

**Comment by yychayitu** [comment]
同问

**Comment by siyilingting** [comment]
我也遇到了该问题，labels是0

**Comment by siyilingting** [comment]
> 我也遇到了该问题，labels是0

问题在于`val.py`中的的[L270-L276](https://github.com/hukaixuan19970627/yolov5_obb/blob/master/val.py#L270)，当模型很差、识别结果全部错误时，‘stats[0].any()==False’， 导致labels数量不被计算，结果自然为0。

yolov5作者其本意是当全部识别错误时，没有必要再计算mAP指标，只是这种写法牵连了ground-truth数量的计算。https://github.com/ultralytics/yolov5/issues/3916#issuecomment-875607211
